File: Projects/Ex56RecyclerView/app/src/main/java/com/c282804/ex56recyclerview/MyAdapter.java
package com.c282804.ex56recyclerview;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;

public class MyAdapter extends RecyclerView.Adapter {

    Context context;
    ArrayList<String> items ;



    public MyAdapter(ArrayList<String> items, Context context) {
        this.context = context;
        this.items = items;

    }

    //재활용할 뷰가 없으면 뷰를 만들어내는 작업을 수행하는 메소드.
    //단, Viewholder객체를 만들어서 리턴
    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

        LayoutInflater inflater = LayoutInflater.from(context);
        View itemView= inflater.inflate(R.layout.recycler_item,parent,false);

        //뷰홀더 객체 생성 및 리턴
        VH holder = new VH(itemView);
        return holder;
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        VH vh = (VH) holder;
        vh.tv.setText(items.get(position));

    }

    //뷰에 값을 연결하는 작업메소드.
    //뷰 설정된 tag 객체 (View holder)를 첫번째 파라미터로 전달해줌.
    @Override
    public int getItemCount() {
        return items.size();
    }
    //이너클래스 : 아이템뷰안에 뷰들의 참조변수를 멤버변수로 가질 클래스
    class VH extends RecyclerView.ViewHolder{
        TextView tv;

        public VH(@NonNull View itemView) {
            super(itemView);
            tv=itemView.findViewById(R.id.tv);

            //리사이클러의 아이템클릭에 반응하는 리스너객체 생성 및 추가
            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    //이 아이템 뷰의 위치 index번호
                    int position= getLayoutPosition();
                    Toast.makeText(context, ""+position, Toast.LENGTH_SHORT).show();
                }
            });
        }
    }
}

File: Projects/Ex91KotlinRecyclerViewApp/app/src/main/java/com/c282804/ex91kotlinrecyclerviewapp/MyAdapter.kt
package com.c282804.ex91kotlinrecyclerviewapp

import android.content.Context
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Adapter
import androidx.recyclerview.widget.RecyclerView
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.recycler_item.view.*



class MyAdapter constructor(var context: Context, val item:ArrayList<Item>) : RecyclerView.Adapter<RecyclerView.ViewHolder>(){
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val layoutInflater=LayoutInflater.from(context)
        val itemView= layoutInflater.inflate(R.layout.recycler_item,parent,false)
        val vh= VH(itemView)
        return vh

    }

    override fun getItemCount(): Int {
        return  item.size
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val vh= holder as VH//코틀린의 형변환 연산자 as

        //현재번째 아이템
        val item= item.get(position)
        vh.itemView.tvTitle.setText(item.title)
        //코틀린에서는 setXXX메소드를 싫어함. 대신 set해서 설정할 값을
        //멤버변수(property:속성)로 설정하는 것을 권장
        vh.itemView.tvMsg.text=item.msg
        //Glide와 같은 역할을 한느 Picasso 라이브러리 사용
        Picasso.get().load(item.img).into(vh.itemView.iv)

        //코틀린에서는 이 위치에서 itemVeiw의 클릭이벤트를 처리를 함
//        vh.itemView.setOnClickListener(object : View.OnClickListener{
//            override fun onClick(v: View?) {
//                //아이템의 상세 정보 화면으로 이동
//            }
//        })
//


    }

    //inner class ViewHolder Class : itemView 안의 뷰들을 관리하는 클래스
    inner  class  VH(itemView: View): RecyclerView.ViewHolder(itemView){
//        val tvTitle=itemView.tvTitle
//        val tvMsg=itemView.tvMsg
//        val iv= itemView.iv


        init {

            //자바에서는 이 생성자에 getLayoutPosition()메소드로
            //클릭한 아이템을 구별했엇음..
            //코틀린에서는 getLayoutPosition()메소드 대신에
            //이 아탑터를 멤버변수로 이미 layoutPosition이 항상존재함
            //근데 layoutPosition 속성값이 언제나 항상-1임
            //그래서 이곳에서 클릭위치를 찾을 수 없음.
            itemView.setOnClickListener(object : View.OnClickListener{
                override fun onClick(v: View?) {
                    //Toast.makeText(context,""+layoutPostition,Toast.LENGTH_SHORT).show()

                    val intent= Intent(context,ItemActivity::class.java)
                    intent.putExtra("title",item.get(layoutPosition).title)
                    intent.putExtra("msg",item.get(layoutPosition).msg)
                    intent.putExtra("img",item.get(layoutPosition).img)

                    context.startActivity(intent)

                }
            })

            //그래서 아이템의 위치정보가 존재하는 bindViewHolder()에서
            //클릭 이벤트 처리
        }

    }
}